Focused text form controls should not get an Interaction Region

etiennesegonzac opened the pull request:
#### a4a7ff9eae45b43802428aa8b520f7c884ad62da
<pre>
Focused text form controls should not get an Interaction Region
<a href="https://bugs.webkit.org/show_bug.cgi?id=258852">https://bugs.webkit.org/show_bug.cgi?id=258852</a>
&lt;rdar://111156944&gt;

Reviewed by Tim Nguyen.

All `HTMLFormControlElement` get an InteractionRegion. For
`HTMLTextFormControlElement`, add a check to make sure they&apos;re not
currently focused.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowNonPointerCursorForElement):
Add the check.

* LayoutTests/interaction-region/text-controls-expected.txt: Added.
* LayoutTests/interaction-region/text-controls.html: Added.
* LayoutTests/interaction-region/text-input-focused-expected.txt: Added.
* LayoutTests/interaction-region/text-input-focused.html: Added.
* LayoutTests/interaction-region/textarea-focused-expected.txt: Added.
* LayoutTests/interaction-region/textarea-focused.html: Added.
New tests covering the change.

Canonical link: <a href="https://commits.webkit.org/265764@main">https://commits.webkit.org/265764@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/636e84ddab5ab6d27eee40ef9fc724011948a80c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12384 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13460 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11253 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11834 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14403 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11995 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14112 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11979 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12808 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/10010 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13882 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10100 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10727 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17857 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11178 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10882 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14052 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11291 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9331 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10606 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2848 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14748 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11147 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -148,6 +148,9 @@ static bool shouldAllowNonPointerCursorForElement(const Element& element)
 #endif
 
+    if (is<HTMLTextFormControlElement>(element))
+        return !element.document().userActionElements().isFocused(element);
```
```suggestion
        return !element.focused();
```

etiennesegonzac replied:
thanks! got confused by the private `isUserActionElementFocused`

nt1m reviewed (approved):
I guess this makes sense since you're already interacting with it when focused.

I do wonder if we should allow cursor: pointer to add an interaction region in this case though.

etiennesegonzac committed:
Focused text form controls should not get an Interaction Region
https://bugs.webkit.org/show_bug.cgi?id=258852
<rdar://111156944>

Reviewed by Tim Nguyen.

All `HTMLFormControlElement` get an InteractionRegion. For
`HTMLTextFormControlElement`, add a check to make sure they're not
currently focused.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::shouldAllowNonPointerCursorForElement):
Add the check.

* LayoutTests/interaction-region/text-controls-expected.txt: Added.
* LayoutTests/interaction-region/text-controls.html: Added.
* LayoutTests/interaction-region/text-input-focused-expected.txt: Added.
* LayoutTests/interaction-region/text-input-focused.html: Added.
* LayoutTests/interaction-region/textarea-focused-expected.txt: Added.
* LayoutTests/interaction-region/textarea-focused.html: Added.
New tests covering the change.

Canonical link: https://commits.webkit.org/265764@main
